sql: add txn diagnostics collection to instrumentation helper

kyle-a-wong opened the pull request:
commit 1/2:

When calling the setup method on instrumentationhelper (ih), ih will
now check if transaction diagnostics should should be collected. Transaction
diagnostics collection uses a prefix based matching system to determine
whether or not transaction diagnostics collection should start, continue,
or stop. This works by creating a list of statement fingerprint ids, that make
up a transaction fingerprint, and comparing the currently executing statement
fingerprint id against this list. If at any point the expected statement fingerprint
id doesn't match, diagnostics collection is short circuited and no longer
collected.

This change also causes transaction diagnostics collection to take precedence
over statement transaction diagnostics, meaning if a transaction diagnostics request
exists that contains a statement that also has a statement diagnostics request,
the transaction diagnostic request gets priority over collecting the bundle.

When a transaction diagnostics request is being fulfilled, statement bundles for
every statement in the transaction are built and collected in memory. Only after
a transaction is finalized and it is determined that the request is fulfilled are all
the statement bundles and the transaction trace stored in the corresponding
system tables.

Resolves: [CRDB-54321](https://cockroachlabs.atlassian.net/browse/CRDB-54321)
Epic: [CRDB-53541](https://cockroachlabs.atlassian.net/browse/CRDB-53541)
Release note: None

----

commit 2/2:
sql: replace statement fingerprint id construction with fingerprint id from instrumentation helper

The instrumentation helper now generates a statement fingerprint id in its
setup function and stores it in its state. Now, when fingerprint ids
are needed for sql stats collection or telemetry logging, the fingerprint
id from the instrumentation helper is used.

Epic: None
Release note: None

kyle-a-wong committed:
sql: add txn diagnostics collection to instrumentation helper

When calling the setup method on instrumentationhelper (ih), ih will
now check if transaction diagnostics should should be collected. Transaction
diagnostics collection uses a prefix based matching system to determine
whether or not transaction diagnostics collection should start, continue,
or stop. This works by creating a list of statement fingerprint ids, that make
up a transaction fingerprint, and comparing the currently executing statement
fingerprint id against this list. If at any point the expected statement fingerprint
id doesn't match, diagnostics collection is short circuited and no longer
collected.

This change also causes transaction diagnostics collection to take precedence
over statement transaction diagnostics, meaning if a transaction diagnostics request
exists that contains a statement that also has a statement diagnostics request,
the transaction diagnostic request gets priority over collecting the bundle.

When a transaction diagnostics request is being fulfilled, statement bundles for
every statement in the transaction are built and collected in memory. Only after
a transaction is finalized and it is determined that the request is fulfilled are all
the statement bundles and the transaction trace stored in the corresponding
system tables.

Resolves: CRDB-54321
Epic: CRDB-53541
Release note: None

dhartunian commented on pkg/sql/instrumentation.go:
```diff
@@ -445,10 +452,17 @@ func (ih *instrumentationHelper) Setup(
+
+		// If no transaction diagnostics are in progress, check for statement-level
+		// diagnostics
+		if !txnHelper.DiagnosticsInProgress() {
```
also, the call above should return a boolean that we can check instead of calling back into `txnHelper`. Again, for the overwhelmingly common case of no txn bundle collection we should do as little work as possible.

kyle-a-wong replied:
done

dhartunian reviewed (commented):
The commit message needs some structure. I think the "core" of all this work that people should understand is the prefix-matching so the commit message should lead with what that is and how it works in a bit more detail.

Can you also discuss how the logic is changing that's in the hot path of the conn_executor? I think that's important because it affects every execution.

kyle-a-wong commented on pkg/sql/instrumentation.go:
```diff
@@ -445,10 +452,17 @@ func (ih *instrumentationHelper) Setup(
-		// IsRedacted will be false when ih.collectBundle is false.
-		ih.explainFlags.RedactValues = ih.explainFlags.RedactValues || ih.diagRequest.IsRedacted()
+		// Handle transaction-level diagnostics
+		ih.handleTransactionDiagnostics(&ex.state, ex.transitionCtx.tracer, stmt, stmtFingerprintId)
```
@dhartunian What do you think about leaving the stmt fingerprint computation in the instrumentation helper and using it here and replacing other places where it is generated with a reference to ih.fingerprintId instead? I see [this PR](https://github.com/cockroachdb/cockroach/pull/124250) that mentions: 
```
Currently we construct the stmt fingerprint id at various places to
pass to the logging function. We only need the statement fingerprint
id for telemetry logging, so this construction is wasteful in other
scenarios. This commit moves the stmt fingerprint id generation to
the logging function.
```
but I'm not sure if this is correct. We construct the fpId [here](https://github.com/cockroachdb/cockroach/blob/2415bcd4b1a154822afc1156b741ead43299d19d/pkg/sql/executor_statement_metrics.go#L194) which is called in every `dispatchToExecutionEngine` call. It seems like we generate a fingerprint id on every stmt execution. We can replace the usage here and where its generated for the `SampledQuery` log to just read from the instrumentation helper, what do you think?

kyle-a-wong committed:
sql: replace statement fingerprint id construction with fingerprint id from instrumentation helper

The instrumentation helper now generates a statement fingerprint id in its
setup function and stores it in its state. Now, when fingerprint ids
are needed for sql stats collection or telemetry logging, the fingerprint
id from the instrumentation helper is used.

Epic: None
Release note: None

kyle-a-wong commented on pkg/sql/txn_instrumentation.go:
```diff
@@ -105,17 +123,20 @@ type txnInstrumentationHelper struct {
-	return h.diagnosticsCollector.requestId != 0
+	return h.diagnosticsCollector.requestId != 0 ||
+		len(h.diagnosticsCollector.stmtBundles) > 0 ||
+		len(h.diagnosticsCollector.stmtsFpsToCapture) > 0
```
done

kyle-a-wong commented on pkg/sql/instrumentation.go:
```diff
@@ -1214,3 +1238,37 @@ func (ih *instrumentationHelper) SetIndexRecommendations(
+) {
+	shouldCollect := false
+
+	if txnState.mu.stmtCount == 1 {
```
Left a comment why we dont need the lock, but as per the documentation on txnstate, we dont need to aquire the lock for reads, since this is running in the same go routine that the con executor is executing on

dhartunian commented on pkg/sql/exec_log.go:
I know somewhere in SQL Stats we recompute fingerprints because the implicitTxn flag may change. Does this matter here? https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlstats/sslocal/sql_stats_ingestor.go#L412

kyle-a-wong replied:
@dhartunian From my understanding, an implicit transaction may be upgraded into an explicit one after the execution of a statement. The sql_stats_ingestor will recompute the fingerprint when flushing a recorded transaction stats object, which is after all the statements in a transaction have been executed. The change i made above is happening when the exec_log is logging the statement during its execution, so i dont think its possible for the implicit txn state to have changed here. Lmk if what i said here is wrong

dhartunian commented on pkg/sql/txn_instrumentation.go:
```diff
@@ -105,17 +123,31 @@ type txnInstrumentationHelper struct {
 
 func (h *txnInstrumentationHelper) Reset() {
 	if h.diagnosticsCollector.requestId != 0 {
+		h.diagnosticsCollector.span.SetRecordingType(tracingpb.RecordingOff)
```
If we're Resetting, shouldn't the span be set to nil so we can get a new one?

kyle-a-wong replied:
@dhartunian we are turning off recording for the span so that nothing is recorded for the remainder of the transaction since we are no longer collecting a transaction diagnostic. When we reset, the diagnostic is reset to the zero value to tell the instrumentation helper that we will no longer collect diagnostics for this transaction and the request is "released" to let a new transaction attempt to fulfill it. We will never need to get a new span because the transaction diagnostics has been aborted. 

when a new instrumentationHelper is created for a new transaction, it will start a new span for that transaction and attach it to the diagnosticsCollector created for that execution.

dhartunian reviewed (commented):
The ASCII diagram is not helpful as-is. I would either ditch it or revise significantly depending on your preference.

A few notes:
- text in boxes is too abbreviated and contains grammatical errors
- silly to mention but the boxes should be...boxes. rounded boxes make the diagram challenging to parse and read since they introduce wacky characters
- the diagram is missing a box for "stop recording if in progress"
- there are arrows to nowhere or...going behind boxes and merging with others (the "no" option from "is txn diagnostic in progress" box, does that merge with the "no" from the "next statement match"?)

yuzefovich reviewed (commented):
> @yuzefovich reviewed 4 of 10 files at r21, 2 of 10 files at r22, 3 of 3 files at r28, 2 of 2 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153296)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @michae2)

___
*[`pkg/sql/instrumentation.go` line 1264 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/153296#-O_RGb8790hzc0Tu2l6K:-O_RGb8790hzc0Tu2l6L:b-gaisie) ([raw file](https://github.com/cockroachdb/cockroach/blob/3062d224566b2d081d71c6440921d4a9571499d0/pkg/sql/instrumentation.go#L1264)):*
> ```Go
> 			// Set txn_state context to the one returned by MaybeStartDiagnostics,
> 			// which contains the trace span started for diagnostics
> 			txnState.Ctx = newCtx
> ```

Modifying `txnState.Ctx` feels a bit fragile, but I don't have a better suggestion.

Some concerns that we want to make sure to be addressed / considered:
- if the old `txnState.Ctx` contains a tracing span, will it be finished or will it be leaked? (I think currently it's the latter since in `txnState.finishSQLTxn` we'll finish the span that we find from the current context.)
- how does txn diagnostics interact with the SET TRACING stmts within a running txn? In `SessionTracing.StartTracing` we're doing something similar to what we're doing here for txn diagnostics, and it feels like we'll break the session tracing. Perhaps we should explicitly fail SET TRACING stmt without putting the txn into erroneous state if we happen to be collecting txn diagnostics? Or do we not expect to ever collect a txn bundle for a txn with SET TRACING stmts?
- `instrumentationHelper.Setup` already returns possibly new `context.Context` object that will be used throughout the execution, and that context is a descendant of the old txn context. Here we're deriving a new txn context without propagating it "out", i.e. I don't think that the context used throughout the execution of the current stmt will be a descendant of the newly-derived txn context, so the current stmt won't be included into the trace. Furthermore, we already added a few other levels of contexts between the old txn one and the current execution one (there are 3 more spots in `execStmtInOpenState` before we call `Setup` where we derive a new context; maybe there are others(?)), so we can't simply return the new txn context to be used as the new execution one.
___
*[`pkg/sql/txn_instrumentation.go` line 150 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/153296#-O_RFRi69DAIZFTABbyI:-O_RFRi69DAIZFTABbyJ:bbht483) ([raw file](https://github.com/cockroachdb/cockroach/blob/3062d224566b2d081d71c6440921d4a9571499d0/pkg/sql/txn_instrumentation.go#L150)):*
> ```Go
> func (h *txnInstrumentationHelper) Reset() {
> 	if h.diagnosticsCollector.requestId != 0 {
> 		h.diagnosticsCollector.span.SetRecordingType(tracingpb.RecordingOff)
> ```

Can `span` be nil here? What if the txn request was redacted?


<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
tftr

bors r+

yuzefovich commented:
Nice! LGTM
